samples/time/timer.cpp

```diff
@@ -0,0 +1,76 @@
+#include <iostream>
+#include <string>
+
+#include <Wink/address.h>
+#include <Wink/machine.h>
+#include <Wink/state.h>
+
+int main(int argc, char **argv) {
+  if (argc < 3) {
+    error() << "Incorrect parameters, expected <spawner> <address> <duration>\n"
+            << std::flush;
+    return -1;
+  }
+
+  Address spawner(argv[1]);
+  Address address(argv[2]);
+  UDPSocket socket;
+  Machine m(spawner, address, "time/Timer", socket);
+
+  std::istringstream iss(argv[3]);
+  int seconds;
+  iss >> seconds;
+  std::time_t start = std::time(nullptr);
+
+  m.AddState(std::make_unique<State>(
+      // State Name
+      "idle",
+      // Parent State
+      "",
+      // On Entry Action
+      []() { info() << "Timer is IDLE\n"
+                    << std::flush; },
+      // On Exit Action
+      []() {},
+      // Receivers
+      std::map<const std::string, Receiver>{
+          {"idle", [&](const Address &sender,
+                       std::istream &args) { m.GotoState("idle"); }},
+          {"start",
+           [&](const Address &sender, std::istream &args) {
+             start = std::time(nullptr);
+             m.GotoState("timing");
+           }},
+          {"stop",
+           [&](const Address &sender, std::istream &args) { m.Exit(); }},
+          {"exit",
+           [&](const Address &sender, std::istream &args) { m.Exit(); }},
+      }));
+
+  m.AddState(std::make_unique<State>(
+      // State Name
+      "timing",
+      // Parent State
+      "",
+      // On Entry Action
+      [&]() { m.SendSelf("update"); },
+      // On Exit Action
+      []() {},
+      // Receivers
+      std::map<const std::string, Receiver>{
+          {"update", [&](const Address &sender, std::istream &args) {
+             const auto now = std::time(nullptr);
+             const auto elapsed = now - start;
+             const auto remaining = seconds - elapsed;
+             info() << "Timer is TIMING: " << remaining << "s left\n"
+                    << std::flush;
+             if (remaining > 0) {
+               m.SendSelf("update"); // Loop
```

Comment:
Fair enough. My original thinking was that this timer was the thread that is spawned to sleep and then send a message, but you're right the first approach wasn't lively and thus could not be cancelled.

This sample could be changed to spawn a thread that sleeps and then sends this timer a message, and if the timer is in the `timing` state it will notify the original spawner, and if it is in the `idle` state (ie if it was cancelled) then it will discard the notification. Though it would make the sample more complex and require low level calls like fork, which I would really rather avoid in sample code that is meant to show how HSMs can make developers lives easier. Maybe this is a bad example and there are better applications of HSMs?

Reply:
It's a good example and shows that SMs force u to think about edge cases. But it also indicates the Wink system needs to provide a "delay" service that you send a message and will receive a response at a future time. I would say a production worthy "delay" service would be a very good example!

Reply:
I was thinking the Timer SM would be the mechanism to get a message after a certain delay, but I agree this wasn't clear so I've added `samples/time/Delay` which spawns a Timer SM so it can automatically exit after 5 seconds.

I see some value in a `SendAfterDelay("msg", 10)` but there are also some drawbacks so generally I prefer exploring new ideas in samples/libraries where they can be quickly iterated upon to find the best solution, before merging that into the framework/language/runtime.

Reply:
After thinking about this for a few days I've come around to your way of thinking.

However, rather than a specific function `SendAfterDelay` which can only be used for sending messages after certain delay, I chose to go with a more generic [`CallAfter`](https://github.com/AletheiaWareLLC/Wink/pull/27) which can be used to do anything after a certain delay.

Reply:
Hmm, the problem using a generic `After` or and using closures is that it you make to easy to make a mistake. For instance calling a SM function or using "captured" SM state, doing can be **VERY bad**, so I caution providing a way to shoot yourself in the foot.

Also, using "deadline" semantics (send a message at a particular time in the future) can sometimes be more natural, you might consider providing both.

Reply:
Ah yeah good point - I actually shot myself in the foot earlier!

I've updated https://github.com/AletheiaWareLLC/Wink/pull/27 to replace `After` with `SendAt` and `SendAfter`